Sources/Sampler/vmc_sampling.cc

```diff
@@ -179,6 +180,307 @@ VectorXcd GradientOfVariance(const Result &result, AbstractMachine &psi,
   MeanOnNodes<>(grad);
   return grad;
 }
-
 }  // namespace vmc
+
+void StepsRange::CheckValid() const {
+  const auto error = [this](const char* expected) {
+    std::ostringstream msg;
+    msg << "invalid steps range: (start=" << start_ << ", end=" << end_
+        << ", step=" << step_ << "); " << expected;
+    throw InvalidInputError{msg.str()};
+  };
+  if (start_ < 0) error("expected start >= 0");
+  if (end_ < 0) error("expected end >= 0");
+  if (step_ <= 0) error("expected step >= 1");
+  if (end_ < start_) error("expected start <= end");
+}
+
+namespace detail {
+// Main loop: iterations specified by `steps` are `record`ed, others are
+// `skip`ped.
+template <class Skip, class Record>
+void LoopV2(StepsRange const& steps, Skip&& skip, Record&& record) {
+  auto i = Index{0};
+  // Skipping [0, 1, ..., start)
+  for (; i < steps.start(); ++i) {
+    skip();
+  }
+  // Record [start]
+  record();
+  for (i += steps.step(); i < steps.end(); i += steps.step()) {
+    for (auto j = Index{1}; j < steps.step(); ++j) {
+      skip();
+    }
+    record();
+  }
+}
+
+void SubtractMean(RowMatrix<Complex>& gradients) {
+  VectorXcd mean = gradients.colwise().mean();
+  assert(mean.size() == gradients.cols());
+  MeanOnNodes<>(mean);
+  gradients.rowwise() -= mean.transpose();
+}
+}  // namespace detail
+
+std::tuple<RowMatrix<double>, Eigen::VectorXcd,
+           nonstd::optional<RowMatrix<Complex>>>
+ComputeSamples(MetropolisLocalV2& sampler, StepsRange const& steps,
+               bool compute_gradients) {
+  sampler.Reset();
+  const auto num_samples = steps.size() * sampler.BatchSize();
+
+  RowMatrix<double> samples(num_samples, sampler.SystemSize());
+  Eigen::VectorXcd values(num_samples);
+  auto gradients =
+      compute_gradients
+          ? nonstd::optional<RowMatrix<Complex>>{nonstd::in_place, num_samples,
+                                                 sampler.Machine().Npar()}
+          : nonstd::nullopt;
+
+  struct Skip {
```

Comment:
Is it really necessary to define specialized structs for very simple operations?